pandas/tests/typing/valid/test_frame.py

```diff
@@ -0,0 +1,853 @@
+"""
+Copyright (c) Virtus Lab sp. z o.o. (Ltd.)
+
+Distributed under the terms of the MIT license.
+
+The full license is in the STUBS_LICENSE file, distributed with this software.
+"""
+# flake8: noqa: F841
+# TODO: many functions need return types annotations for pyright
+# to run with reportGeneralTypeIssues = true
+import io
+import os
+from pathlib import Path
+import tempfile
+from typing import (
+    Any,
+    Iterable,
+    List,
+    Tuple,
+)
+
+import numpy as np
+
+import pandas as pd
+from pandas.util import _test_decorators as td
+
+
+def test_types_init() -> None:
+    pd.DataFrame(data={"col1": [1, 2], "col2": [3, 4]})
+    pd.DataFrame(data={"col1": [1, 2], "col2": [3, 4]}, index=[2, 1])
+    pd.DataFrame(data=[1, 2, 3, 4], dtype=np.int8)
+    pd.DataFrame(
+        np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]]),
+        columns=["a", "b", "c"],
+        dtype=np.int8,
+        copy=True,
+    )
+
+
+def test_types_to_csv() -> None:
+    df = pd.DataFrame(data={"col1": [1, 2], "col2": [3, 4]})
+    # error: Incompatible types in assignment (expression has type "Optional[str]",
+    # variable has type "str")
+    csv_df: str = df.to_csv()  # type: ignore[assignment]
+
+    # NamedTemporaryFile cannot be used with delete=True on Windows
+    # see https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile
+    try:
+        with tempfile.NamedTemporaryFile(delete=False) as file:
```

Comment:
I think it might be better to avoid using `tempfile.*TemporaryFile` if the file handle is not actually used.

If you want to provide the pandas API with a str/Path, I would suggest using:

```py
import pandas._testing as tm

with tm.ensure_clean() as path:
    # path has no extension
    pd.read_csv(path)

with tm.ensure_clean(".some_suffix") as path:
    # path ends with ".some_suffix"
    pd.read_csv(path)
```

Might be good to grep for all `TemporaryFile` to change this everywhere.

Reply:
Sorry, one more comment: I think that ensure_clean is not typed: you might need to explicitly convert it to a string so that mypy knows it is `str` and not `Any`. Might be worth to first use `reveal_type(path)` maybe it is already being picked up as `str`.